### ISSUE TITLE
sess-02

### ISSUE THREAD

**Issue by jordandekraker** [khanlab/hippunfold#46]
Just trying out the pipeline with some multi-session T1w data, I believe the desired behaviour is that each session should run independently (eg. for a longitudinal study). Currently, this returns the following error:
```
(venv) jordand@valentine:/data/mica3/jordand$ hippunfold ../BIDS_MIC/rawdata/ BIDS_MIC_raw participant --participant_label HC001 --modality T1w --use-singularity --cores all
Building DAG of jobs...
InputFunctionException in line 4 of /data/mica3/jordand/hippunfold/hippunfold/workflow/rules/preproc_t1.smk:
Error:
  IndexError: list index out of range
Wildcards:
  subject=HC001
  session=02
Traceback:
  File "/data/mica3/jordand/hippunfold/hippunfold/workflow/rules/preproc_t1.smk", line 7, in <lambda>
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/__init__.py", line 204, in filter_list
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/__init__.py", line 204, in <dictcomp>
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/__init__.py", line 204, in <listcomp>
WARNING: more than one snakemake filename for T1w, taking the first
  To correct this, use the --filter_T1w option to narrow the search
{'/data/mica3/BIDS_MIC/rawdata/sub-{subject}/ses-{session}/anat/sub-{subject}_ses-{session}_T1w.nii.gz', '/data/mica3/BIDS_MIC/rawdata/sub-{subject}/ses-{session}/anat/sub-{subject}_ses-{session}_run-{run}_T1w.nii.gz'}

Traceback (most recent call last):
  File "/data/mica3/jordand/venv/bin/hippunfold", line 33, in <module>
    sys.exit(load_entry_point('hippunfold', 'console_scripts', 'hippunfold')())
  File "/data/mica3/jordand/hippunfold/hippunfold/run.py", line 14, in main
    app.run_snakemake()
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/app.py", line 209, in run_snakemake
    run(snakemake_cmd)
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/app.py", line 41, in run
    raise Exception("Non zero return code: %d"%process.returncode)
```

**Comment by jordandekraker** [comment]
This has still been a bit tricky since session-01 has only 1 run (and therefore no run spec in the filename), while session-02 has 2 runs. I tried a few different filters (eg. --filter_T1w session=02 --filter_T1w run=1) and couldn't get it to run. I suppose this really is an issue with the PyBIDS grabber and the documentation wasn't super clear to me there. Still, not a huge deal, I can still move forward.

**Comment by jordandekraker** [comment]
Still something funny happening, not sure why but it doesn't seem to recognize 'session' or 'ses'
```
(venv) jordand@valentine:/data/mica3/jordand$ hippunfold ../BIDS_MIC/rawdata/ BIDS_MIC_raw2 participant --participant_label HC001 --modality T1w --filter_T1w session=02 run=1 --use-singularity --cores all
Building DAG of jobs...
InputFunctionException in line 4 of /data/mica3/jordand/hippunfold/hippunfold/workflow/rules/preproc_t1.smk:
Error:
  IndexError: list index out of range
Wildcards:
  subject=HC001
  session=01
Traceback:
  File "/data/mica3/jordand/hippunfold/hippunfold/workflow/rules/preproc_t1.smk", line 7, in <lambda>

Traceback (most recent call last):
  File "/data/mica3/jordand/venv/bin/hippunfold", line 33, in <module>
    sys.exit(load_entry_point('hippunfold', 'console_scripts', 'hippunfold')())
  File "/data/mica3/jordand/hippunfold/hippunfold/run.py", line 14, in main
    app.run_snakemake()
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/app.py", line 209, in run_snakemake
    run(snakemake_cmd)
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/app.py", line 41, in run
    raise Exception("Non zero return code: %d"%process.returncode)
Exception: Non zero return code: 1
(venv) jordand@valentine:/data/mica3/jordand$ hippunfold ../BIDS_MIC/rawdata/ BIDS_MIC_raw2 participant --participant_label HC001 --modality T1w --filter_T1w ses=02 run=1 --use-singularity --cores all
KeyError in line 56 of /data/mica3/jordand/hippunfold/hippunfold/workflow/Snakefile:
'T1w'
  File "/data/mica3/jordand/hippunfold/hippunfold/workflow/Snakefile", line 143, in <module>
  File "/data/mica3/jordand/hippunfold/hippunfold/workflow/Snakefile", line 56, in get_final_specs
WARNING: no images found for T1w

Traceback (most recent call last):
  File "/data/mica3/jordand/venv/bin/hippunfold", line 33, in <module>
    sys.exit(load_entry_point('hippunfold', 'console_scripts', 'hippunfold')())
  File "/data/mica3/jordand/hippunfold/hippunfold/run.py", line 14, in main
    app.run_snakemake()
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/app.py", line 209, in run_snakemake
    run(snakemake_cmd)
  File "/data/mica3/jordand/venv/lib/python3.7/site-packages/snakebids/app.py", line 41, in run
    raise Exception("Non zero return code: %d"%process.returncode)
Exception: Non zero return code: 1
```

**Comment by akhanf** [comment]
can you paste the folder structure for a subject?

**Comment by jordandekraker** [comment]
omitting all non anat/ folders:
```
../BIDS_MIC/rawdata/sub-HC001/
├── ses-01
│   ├── anat
│   │   ├── sub-HC001_ses-01_acq-inv1_T1map.json
│   │   ├── sub-HC001_ses-01_acq-inv1_T1map.nii.gz
│   │   ├── sub-HC001_ses-01_acq-inv2_T1map.json
│   │   ├── sub-HC001_ses-01_acq-inv2_T1map.nii.gz
│   │   ├── sub-HC001_ses-01_acq-mp2rage_T1map.json
│   │   ├── sub-HC001_ses-01_acq-mp2rage_T1map.nii.gz
│   │   ├── sub-HC001_ses-01_acq-uni_T1map.json
│   │   ├── sub-HC001_ses-01_acq-uni_T1map.nii.gz
│   │   ├── sub-HC001_ses-01_T1w.json
│   │   └── sub-HC001_ses-01_T1w.nii.gz
└── ses-02
    ├── anat
    │   ├── sub-HC001_ses-02_acq-inv1_T1map.json
    │   ├── sub-HC001_ses-02_acq-inv1_T1map.nii.gz
    │   ├── sub-HC001_ses-02_acq-inv2_T1map.json
    │   ├── sub-HC001_ses-02_acq-inv2_T1map.nii.gz
    │   ├── sub-HC001_ses-02_acq-mp2rage_T1map.json
    │   ├── sub-HC001_ses-02_acq-mp2rage_T1map.nii.gz
    │   ├── sub-HC001_ses-02_acq-mp2rage_T1map_ROI1.nii.gz
    │   ├── sub-HC001_ses-02_acq-uni_T1map.json
    │   ├── sub-HC001_ses-02_acq-uni_T1map.nii.gz
    │   ├── sub-HC001_ses-02_run-1_T1w.json
    │   ├── sub-HC001_ses-02_run-1_T1w.nii.gz
    │   ├── sub-HC001_ses-02_run-2_T1w.json
    │   └── sub-HC001_ses-02_run-2_T1w.nii.gz
```

**Comment by akhanf** [comment]
thanks -- I am able to reproduce the error now - will take a look at it

**Comment by akhanf** [comment]
I see what is happening -- in snakebids, the subjects/sessions are always pulled from the dataset as a whole, regardless of what filters are being used.. (so setting `--filter_T1w session=01` doesn't end up removing 02 from the list of sessions..)

With subjects we don't have the problem since the --participant_label overrides it, but there isn't a way to override the list of sessions right now..

So I think the fix needs to happen in snakebids -- will make an issue there -- 

Another hacky workaround to try in the meantime might be to use .bidsignore to ignore the session??

**Comment by jordandekraker** [comment]
OK sounds good! I'll keep an eye on snakebids. 

I've already implemented a hacky workaround for now so no rush :)

**Comment by akhanf** [comment]
ok! I'll actually leave this open until the fix is in -- actually the  snakebids 0.3.0 could also deal with this since can skip the bids parsing altogether..

**Comment by jordandekraker** [comment]
Beautiful implementation, Ali! Seems to be working well on my end, just a couple of minor issues

bug: when I load one of these resampled surfaces in Freeview, it crashes when i pan to the edge of the surface (but shows up correctly in the 3D display). I thought this might have to do with the surface exceeding the bounding box maybe? Might just be a Freeview thing though. 

suggestion: maybe default number of vertices should be 2k. I think many neocortical atlases have on the order of 32k vertices (eg. conte69 as Boris is using), so 2k would probably be roughly proportional.

suggestion: specifying --output_density 32k could use the original 126x254 mesh

**Comment by akhanf** [comment]
@jordandekraker was this comment supposed to be on Issue #53 instead?

**Comment by jordandekraker** [comment]
oops! yes it was, let me just delete and repost there

**Comment by akhanf** [comment]
I think this can be closed, but is related to issue #80 (requires use of `--path_*` options)

**Comment by jordandekraker** [comment]
^ Agreed. See https://github.com/khanlab/hippunfold/issues/80 for any issues with this in the future
